sdk/appconfiguration/dirs.proj

```diff
@@ -0,0 +1,5 @@
+<Project Sdk="Microsoft.Build.Traversal">
+    <ItemGroup>
+        <ProjectReference Include="**\*.csproj" />
```

Comment:
We may want to consider a generic dirs.proj file under eng that allows you to pass in the service directory for the traversal. That would help eliminate the need to create these in ever service folder.

Reply:
So do an import for a eng\dirs.proj and but set a property that specifies the service directory?

Reply:
Yes that is the idea I was thinking about. I'm not sure if it is better or not until I see it in use. 

You would need to pass the `/p:ServiceDirectory=$(SerivceDirectory)` property to all the places you pass the project path.

Reply:
If we think these traversals will only be used by the engineering system then this approach might be reasonable, but if we feel people working in the repo might use them then your current approach is probably better.

Another thing that makes using a generic one inside of the eng folder better is that it will remove ambiguity in the service directories if you build from the command line. Today there is just a solution file so if you call msbuild or dotnet build in the service directory you don't need to pass the project or sln file to it, but as soon as you add these project people will need to specify either the solution or the dirs.proj file when calling build commands in the repo.

Reply:
I think you would end up with a bunch of files that look like this:

```xml
<Project Sdk="Microsoft.Build.Traversal">
  <PropertyGroup>
    <ServicePath>sdk/servicebus</ServicePath>
  </PropertyGroup>
  <Import Project="..\..\eng\dirs.props" />
</Project>
```

Versus, what we currently have:

```xml
<Project Sdk="Microsoft.Build.Traversal">
  <ItemGroup>
    <ProjectReference Include="**\*.csproj" />
  </ItemGroup>
</Project>
```

I'm not sure it adds much. Even though these files are identical from project to project their presence is largely necessary and if there is any customization of what gets built/excluded, that configuration is localized to the service folder (similar to how you've data plane works but only on a case by case basis).

Reply:
> Today there is just a solution file so if you call msbuild or dotnet build in the service directory you don't need to pass the project or sln file to it, but as soon as you add these project people will need to specify either the solution or the dirs.proj file when calling build commands in the repo.

I was going to bring this up. I think that there is _some_ inconsistency today around where solution files in the service folders. It would be great to get them all consistent.

Reply:
I was imagining something like:

eng/service.proj
```
<Project Sdk="Microsoft.Build.Traversal">
  <ProjectReference Include="sdk\$(ServiceName)\**\*.csproj" />
</Project>
```

Then in our ci yaml template we would essentially have a project path that looks like:
```
eng\service.proj /p:ServiceName=$(ServiceName)
```

We could leverage the injection hole to set the property or we could make it a first class parameter :)

Reply:
> if there is any customization of what gets built/excluded, that configuration is localized to the service folder (similar to how you've data plane works but only on a case by case basis).

We could still allow customization via a common props file we import from the service directory if it exists.

Reply:
>I was going to bring this up. I think that there is some inconsistency today around where solution files in the service folders. It would be great to get them all consistent.

Some services that have multiple libraries want to have a solution that contains all so the general guidance is to put the solution at the scope that incorporates all the things they want to include. Of course that currently breaks for things like project references for Azure.Core but we have to draw the line somewhere.

Reply:
OK - I'm on the fence on this one :) I can see having a single ```service.proj``` file in the ```eng/``` path has less duplication but also leads to less flexibility if folks want to exclude things. It makes about 10 characters worth of difference in the YAML files ;)

Part of me likes having the ```dirs.proj``` from a flexibility perspective but can't think of a reason to keep them over the centralized approach at the moment.

<img src="https://media.giphy.com/media/jPAdK8Nfzzwt2/giphy.gif">

Reply:
I don't like the ambiguity aspect but we can actually solve that a different way as well by simply changing the file extension of dirs.proj to be something like dirs.build or something like that. msbuild doesn't care about the file extension (although the traversal extension might but I'd have to test it). 

Personally unless we feel users of the repo would actually want to use the dirs.proj files I'd go with the central one for now as it has less moving parts and introduces less files to maintain. If we find that there is other scenarios later we can adjust.